staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -121,6 +122,20 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	case <-after:
 		postTimeoutFn()
 		tw.timeout(err)
+
+		// resultCh needs to have a reader, since the function doing
+		// the work needs to send to it.
+		go func() {
+			res := <-resultCh
```

Comment:
Does this need to be non-blocking? For example, `res, ok := <-resultCh`.

The other thing that could be done here is to have the channel be buffered to a length of 1. That way, the writer just writes and goes away if it returns and you don't need this extra goroutine.

Reply:
Yes, it has to be blocking, otherwise it could exit and leave the channel with a writer but no reader.

Buffering it is a solution, but in that case the stack trace would vanish into the ether and we really want it to show up in the logs.

Reply:
Ok, I see. You want a panic or a stack trace if the goroutine returns later.

Reply:
One more possible approach: try to send in the defer, fail if the listener is gone and print the `utilruntime.HandleError(t)` if that fails:

```
select {
case resultCh <- err:
default:
  utilruntime.HandleError(err)
}
```

Reply:
That was actually my first fix, but it has a race-- we can end up sending before there is a reader, and then a successful request will end up appearing to time out.

Reply:
I thought of an alternative that avoids spawning a new goroutine, and avoids the race:

```go
timedOutCh := make(chan struct{})

...

select {
case resultCh <- err:
  return
case <-timedOutCh:
  if res != nil {
    // if we know we timed out and have no receiver, log if non-nil
    ...
  }
}

...

select {
case err := <-resultCh:
  ...
case <-after:
  close(timedOutCh) // indicate we timed out to unblock the sending goroutine
  ...
}
```

Reply:
Hm, I'd want to make a test as I'm not 100% sure analytically that that mechanism is race free.

The current solution works pretty clearly, and is easier to cherry pick.

Starting a goroutine isn't actually very heavyweight.

We can lose this second goroutine in a follow-up, but I think we'd be better served by putting our effort into *actually making the worker goroutine exit*.